File: day7/src/main.py
# Since rust is too hard lmao, let's do it in python

# Read in the file and create a list of the lines
with open("input.txt", "r") as f:
    lines = f.readlines()

class FileSystem :
    def __init__(self, root):
        self.root = root
        self.current_dir = root

class Folder:
    def __init__(self, name):
        self.name = name
        self.parent = None
        self.files = []
        self.folders = []
        self.size = 0

    def set_parent(self, parent):
        self.parent = parent

    def add_file(self, file):
        self.files.append(file)
        self.size += int(file.size)

    def add_folder(self, folder):
        self.folders.append(folder)
        self.size += int(folder.size)

class File:
    def __init__(self, name, size):
        self.name = name
        self.size = size

# Create the root folder
root = Folder("root")

# Create the file system
fs = FileSystem(root)

# Traverse through the command list and build the file system
for line in lines:
    # Print the line
    print(line, end="")

    # Split the line into the command and the path
    cmd = line.split(" ")

    # Strip newlines
    for i, c in enumerate(cmd):
        cmd[i] = c.strip()

    if cmd[1] == "ls":
        continue
    elif cmd[1] == "cd" and cmd [2] != "..":
        # Check if folder exists
        found = False
        for folder in fs.current_dir.folders:
            if folder.name == cmd[2]:
                fs.current_dir = folder
                found = True
                break
        if not found:
            # Create the folder
            new_folder = Folder(cmd[2])
            new_folder.set_parent(fs.current_dir)
            fs.current_dir.add_folder(new_folder)
            fs.current_dir = new_folder
    elif cmd[1] == "cd" and cmd[2] == "..":
        # Go up a directory
        fs.current_dir = fs.current_dir.parent
    elif cmd[0] == "dir":
        # Add the folder to the current dir
        new_folder = Folder(cmd[1])
        new_folder.set_parent(fs.current_dir)
        fs.current_dir.add_folder(new_folder)
    else:
        # Add the file to the current dir
        new_file = File(cmd[1], cmd[0])
        fs.current_dir.add_file(new_file)

# Traverse through the file system and updated the sizes to include the size of the subfolders
def update_sizes(folder):
    if folder.folders == []:
        return
    for subfolder in folder.folders:
        update_sizes(subfolder)
        folder.size += subfolder.size

update_sizes(root)

print("-" * 80)

# Print the file system
def print_fs(folder, indent):
    for files in folder.files:
        print((" " * indent)+ "F " + files.name + " " + str(files.size))
    if folder.folders == []:
        return
    for subfolder in folder.folders:
        print((" " * indent) + "D " + subfolder.name + " " + str(subfolder.size))
        print_fs(subfolder, indent + 1)

print_fs(root, 0)

# Return a list of all directories under a certain size
def get_dirs_under_size(folder, size):
    dirs = []
    if folder.folders == []:
        return dirs
    for subfolder in folder.folders:
        if subfolder.size <= size:
            dirs.append(subfolder)
        dirs += get_dirs_under_size(subfolder, size)
    return dirs

dirs_100k = get_dirs_under_size(root, 100000)

print("-" * 80)

# Print the directories under 100k and their sizes
for dir in dirs_100k:
    print("D " + dir.name + " " + str(dir.size))

# Sum the sizes of the directories
total_size = 0
for dir in dirs_100k:
    total_size += dir.size

print("Total size of directories under 100k: " + str(total_size))

# Part 2
max_fs_size = 70000000
free_space_needed = 30000000
target_space = max_fs_size - free_space_needed
amount_to_delete = fs.root.size - target_space

# Print the amount to delete
print(f"{fs.root.size} - {free_space_needed} = {amount_to_delete}")

# Get the directories over amount_to_delete
def get_dirs_over_size(folder, size):
    dirs = []
    if folder.folders == []:
        return dirs
    for subfolder in folder.folders:
        if subfolder.size >= size:
            dirs.append(subfolder)
        dirs += get_dirs_over_size(subfolder, size)
    return dirs

dirs_over = get_dirs_over_size(root, amount_to_delete)

# Print the directories over the limit
for dir in dirs_over:
    print("D " + dir.name + " " + str(dir.size))

# Sort the directories by size
dirs_over.sort(key=lambda x: x.size, reverse=True)

# Print the size of smallest directory
print("Smallest directory over the limit: " + str(dirs_over[-1].size))

File: day7/src/main.rs
use std::path::PathBuf;

fn main() {
    let mut input_path = PathBuf::new();
    input_path.push(std::env::current_dir().unwrap());
    input_path.push("src/input.txt");

    // Take in the input.txt file and create a vector where each element is a string of the line
    let input_string = match std::fs::read_to_string(input_path) {
        Ok(s) => s,
        Err(e) => {
            println!("Error reading file: {}", e);
            std::process::exit(1);
        }
    };

    let file_vec: Vec<&str> = input_string.lines().collect();

    // Print the file
    for line in &file_vec {
        println!("{}", line);
    }

    pub struct FileSystem {
        root: Box<Directory>,
        current_dir: Box<Directory>,
    }

    pub struct File {
        name: String,
        size: u32,
    }

    pub struct Directory {
        name: String,
        parent: Option<Box<Directory>>,
        files: Option<Box<Vec<File>>>,
        folders: Option<Box<Vec<Directory>>>,
        size: u32,
    }
    impl FileSystem {
        pub fn new(root: Box<Directory>) -> FileSystem {
            FileSystem {
                root: root,
                current_dir: root,
            }
        }
        pub fn get_root(&self) -> &Box<Directory> {
            &self.root
        }
        pub fn get_current_dir(&self) -> &Box<Directory> {
            &self.current_dir
        }
        pub fn set_root(&mut self, root: Directory) {
            self.root = Box::new(root);
        }
        pub fn set_current_dir(&mut self, current_dir: Directory) {
            self.current_dir = Box::new(current_dir);
        }
    }
    impl File {
        pub fn new(name: &str, size: u32) -> File {
            File {
                name: name.to_string(),
                size: size,
            }
        }
    }
    impl Directory {
        pub fn new(name: &str) -> Directory {
            Directory {
                name: name.to_string(),
                parent: None,
                files: None,
                folders: None,
                size: 0,
            }
        }
        pub fn get_name(&self) -> &str {
            &self.name
        }
        pub fn get_parent(&self) -> Option<&Box<Directory>> {
            self.parent.as_ref()
        }
        pub fn get_files(&self) -> Option<&Box<Vec<File>>> {
            self.files.as_ref()
        }
        pub fn get_folders(&self) -> Option<&Box<Vec<Directory>>> {
            self.folders.as_ref()
        }
        pub fn get_size(&self) -> u32 {
            self.size
        }
        pub fn is_parent_some(&self) -> bool {
            self.parent.is_some()
        }
        pub fn is_files_some(&self) -> bool {
            self.files.is_some()
        }
        pub fn is_folders_some(&self) -> bool {
            self.folders.is_some()
        }
        pub fn does_file_exist(&self, name: &str) -> bool {
            if self.files.is_some() {
                for file in self.files.as_ref().unwrap().iter() {
                    if file.name == name {
                        return true;
                    }
                }
            }
            false
        }
        pub fn does_folder_exist(&self, name: &str) -> bool {
            if self.folders.is_some() {
                for folder in self.folders.as_ref().unwrap().iter() {
                    if folder.get_name() == name {
                        return true;
                    }
                }
            }
            false
        }
        pub fn set_parent(&mut self, parent: Directory) {
            self.parent = Some(Box::new(parent));
        }
        pub fn add_file(&mut self, file: File) {
            if self.files.is_none() {
                self.files = Some(Box::new(Vec::new()));
            }
            self.files.as_mut().unwrap().push(file);
        }
        pub fn add_folder(&mut self, folder: Directory) {
            if self.folders.is_none() {
                self.folders = Some(Box::new(Vec::new()));
            }
            self.folders.as_mut().unwrap().push(folder);
        }
        pub fn update_size(&mut self) {
            let mut size = 0;
            if self.files.is_some() {
                for file in self.files.as_mut().unwrap().iter() {
                    size += file.size;
                }
            }
            if self.folders.is_some() {
                for folder in self.folders.as_mut().unwrap().iter() {
                    size += folder.size;
                }
            }
            self.size = size;
        }
    }

    // Go through the command file and create the file system
    let mut root = Box::new(Directory::new("root"));
    let mut file_system = FileSystem::new(root);

    // Go through the command file and create the directory structure
    for line in file_vec {
        let line_vec: Vec<&str> = line.split_whitespace().collect();
        if (line_vec[1] == "cd" && line_vec[2] != "..") {
            // Check if the current directory has any folders
            if file_system.current_dir.is_folders_some() {
                // TODO
            }
        }
    }





    // // Go through the command file and create the directory structure
    // let mut root = Directory::new("root");
    // let mut current_dir = root;
    // for line in file_vec {
    //     let line_vec: Vec<&str> = line.split_whitespace().collect();
    //     if (line_vec[1] == "cd" && line_vec[2] != "..") {
    //         // Check if the folder exists
    //         let mut found = false;
    //         if current_dir.get_folders().is_some() {
    //             for folder in current_dir.get_folders().as_mut().unwrap().iter() {
    //                 if folder.get_name() == &line_vec[2] {
    //                     current_dir = *folder;
    //                     found = true;
    //                     break;
    //                 }
    //             }
    //         }
    //         if !found {
    //             let mut new_dir = Directory::new(line_vec[2]);
    //             new_dir.set_parent(current_dir);
    //             current_dir.add_folder(new_dir);
    //             current_dir = *current_dir.get_folders().as_mut().unwrap().last().unwrap();
    //         }
    //     }
    //     else if (line_vec[1] == "cd" && line_vec[2] == "..") {
    //         current_dir = *current_dir.get_parent().unwrap();
    //     }
    //     else if (line_vec[0] == "ls") {
    //         continue;
    //     }
    //     else if (line_vec[0] == "dir") {
    //         let mut new_dir = Directory::new(line_vec[1]);
    //         new_dir.set_parent(current_dir);
    //         current_dir.add_folder(new_dir);
    //     }
    //     else {
    //         let mut new_file = File {
    //             name: line_vec[1].to_string(),
    //             size: line_vec[2].parse::<u32>().unwrap(),
    //         };
    //         current_dir.add_file(new_file);
    //     }
    // }

    // // Update the size of each directory
    // let mut stack = Vec::new();
    // stack.push(root);
    // while !stack.is_empty() {
    //     let mut current_dir = stack.pop().unwrap();
    //     current_dir.update_size();
    //     if current_dir.get_folders().is_some() {
    //         for folder in current_dir.get_folders().as_mut().unwrap().iter() {
    //             stack.push(*folder);
    //         }
    //     }
    // }

    // // Print the size of each directory
    // let mut stack = Vec::new();
    // stack.push(root);
    // while !stack.is_empty() {
    //     let mut current_dir = stack.pop().unwrap();
    //     println!("{} {}", current_dir.get_name(), current_dir.get_size());
    //     if current_dir.get_folders().is_some() {
    //         for folder in current_dir.get_folders().as_mut().unwrap().iter() {
    //             stack.push(*folder);
    //         }
    //     }
    // }
}
